### PR TITLE
Add auto-configuration for caching 20.10

### PR DESCRIPTION
**Pull request by hayley-jean**
Added: auto configuration for stream cache, reader threads and worker threads.
Changed: gRPC settings

Merge in performance changes for 20.10.3
This PR contains most of the changes in https://github.com/EventStore/EventStore/pull/2902:

Server improvements:
* Auto-size LRU Cache based on the amount of free memory at startup.
* Auto-configure the number of reader threads based on free memory at startup.
* Auto-configure the worker threads based on the reader thread count

Test Client improvements:
* Add gRPC write flood command
* Add TCP ClientAPI write flood command (as opposed to raw TCP)
* Document Test Client Public methods

**Note**: The commits in this PR need to be preserved as commit b4d0bc8a125d16a44f23c392fea20f2aba73ba23 has been tagged for 20.10.3.

### CODE REVIEW

File: src/EventStore.Core/ClusterVNodeStartup.cs [hunk 1]
```diff
@@ -154,7 +155,12 @@ IServiceProvider IStartup.ConfigureServices(IServiceCollection services) => Conf
 						.AddGrpc()
 						.AddServiceOptions<Streams>(options =>
 							options.MaxReceiveMessageSize = TFConsts.EffectiveMaxLogRecordSize)
-						.Services,
+						.Services.Configure<KestrelServerOptions>(options => {
+							options.Limits.MaxConcurrentConnections = 5000;
+							options.Limits.MaxConcurrentUpgradedConnections = 5000;
+							options.Limits.Http2.InitialConnectionWindowSize = 131072 * 1024;
+							options.Limits.Http2.InitialStreamWindowSize = 98304 * 1024;
+						}),
```
Comment (jageall) [inline src/EventStore.Core/ClusterVNodeStartup.cs:163]: These need to be moved to the hostbuilder to have an effect

**Review by jageall — DISMISSED** [review]

File: src/EventStore.Core/ClusterVNodeStartup.cs (hunk 1 above)
Comment (jageall) [reply src/EventStore.Core/ClusterVNodeStartup.cs:163]: They should really have configuration options too

Comment (condron) [reply src/EventStore.Core/ClusterVNodeStartup.cs:163]: not sure they should have cong. options

Comment (hayley-jean) [reply src/EventStore.Core/ClusterVNodeStartup.cs]: I've moved the settings and added configuration options in the same place as the keepAlive settings

**Comment by hayley-jean** [comment]
@ahjohannessen We may decide to go with the approach in PR https://github.com/EventStore/EventStore/pull/2949 for configuring kestrel instead of providing options within Event Store itself. If you have any suggestions for a good default for `MaxConcurrentStreams`, could you comment on that PR please?

**Comment by timothycoleman** [comment]
Is it right that the kestrel config changes aren't in the back port? If so I think this is all set

**Review by thefringeninja — APPROVED** [review]

**Comment by hayley-jean** [comment]
> Is it right that the kestrel config changes aren't in the back port? If so I think this is all set

I'll open a follow-up PR once this is merged with the kestrel settings changes

**Review by timothycoleman — APPROVED** [review]
looks good 👍

**Pull request merged**